File: anthropic/java/anthropic/StringTokenizer.java
package anthropic;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/** Longest-match tokenizer (from string_tokenizer.py). */
public class StringTokenizer {

    static class Trie {
        Map<Character, Trie> children = new HashMap<>();
        String id;
    }

    public List<String> tokenize(String text, List<String> dictionary) {
        Trie root = buildTrie(dictionary);
        List<String> result = new ArrayList<>();
        int i = 0;
        while (i < text.length()) {
            Trie node = root;
            String bestId = null;
            int bestEnd = i;
            int j = i;
            while (j < text.length()) {
                char c = text.charAt(j);
                if (!node.children.containsKey(c)) break;
                node = node.children.get(c);
                j++;
                if (node.id != null) {
                    bestId = node.id;
                    bestEnd = j;
                }
            }
            if (bestId != null) {
                result.add(bestId);
                i = bestEnd;
            } else {
                result.add(String.valueOf(text.charAt(i)));
                i++;
            }
        }
        return result;
    }

    Trie buildTrie(List<String> dictionary) {
        Trie root = new Trie();
        for (String entry : dictionary) {
            int colon = entry.indexOf(':');
            if (colon < 0) continue;
            String key = entry.substring(0, colon);
            String val = entry.substring(colon + 1);
            Trie node = root;
            for (int i = 0; i < key.length(); i++) {
                char c = key.charAt(i);
                node.children.putIfAbsent(c, new Trie());
                node = node.children.get(c);
            }
            node.id = val;
        }
        return root;
    }

    public static void main(String[] args) {
        StringTokenizer s = new StringTokenizer();
        assertEq(
                s.tokenize(
                        "applepiepear",
                        Arrays.asList("app:10", "apple:20", "pie:30")),
                Arrays.asList("20", "30", "p", "e", "a", "r"));
        assertEq(
                s.tokenize("acdebe", Arrays.asList("a:1", "b:2", "cd:3")),
                Arrays.asList("1", "3", "e", "2", "e"));
        System.out.println("StringTokenizer: ok");
    }

    static void assertEq(List<String> a, List<String> b) {
        if (!a.equals(b)) throw new AssertionError(a + " != " + b);
    }
}

File: anthropic/python/string_tokenizer.py
# Example 1:

# Input: text = "applepiepear", dictionary = ["app:10", "apple:20", "pie:30"]
# Output: ["20", "30", "p", "e", "a", "r"]
# Explanation: At the start, "apple" is matched instead of the shorter "app". Then "pie" is matched. The remaining characters do not match any token and are kept as literals.

# Example 2:

# Input: text = "acdebe", dictionary = ["a:1", "b:2", "cd:3"]
# Output: ["1", "3", "e", "2", "e"]

# Example 3:

# Input: text = "programmingprogrampropro", dictionary = ["pro:1", "program:2", "programming:3", "gram:4", "ming:5", "pr:6", "og:7"]
# Output: ["3", "2", "1", "1"]

from typing import List

class Trie:
    def __init__(self):
        self.children = {}
        self.id = None

class Solution:
    def tokenize(self, text: str, dictionary: List[str]) -> List[str]:
        root = self.buildTrie(dictionary)
        result = []

        i = 0
        while i < len(text):
            # Try to find longest match starting at position i
            node = root
            bestId = None
            bestEnd = i

            j = i
            while j < len(text):
                c = text[j]
                if c not in node.children:
                    break
                node = node.children[c]
                j = j + 1

                # Track the last terminal node we've seen
                if node.id is not None:
                    bestId = node.id
                    bestEnd = j

            if bestId is not None:
                result.append(bestId)
                i = bestEnd
            else:
                # No match found, emit literal character
                result.append(str(text[i]))
                i = i + 1

        return result


    def buildTrie(self, dictionary: List[str]) -> Trie:
        root = Trie()

        for entry in dictionary:
            colonIndex = entry.find(':')
            if colonIndex == -1:
                continue

            key = entry[:colonIndex]
            val = entry[colonIndex + 1:]

            node = root
            for i in range(len(key)):
                c = key[i]
                if c not in node.children:
                    node.children[c] = Trie()
                node = node.children[c]
            # Last wins if duplicate keys
            node.id = val

        return root


def _run_tests() -> None:
    s = Solution()

    assert s.tokenize(
        "applepiepear",
        ["app:10", "apple:20", "pie:30"],
    ) == ["20", "30", "p", "e", "a", "r"]

    assert s.tokenize("acdebe", ["a:1", "b:2", "cd:3"]) == ["1", "3", "e", "2", "e"]

    assert s.tokenize(
        "programmingprogrampropro",
        [
            "pro:1",
            "program:2",
            "programming:3",
            "gram:4",
            "ming:5",
            "pr:6",
            "og:7",
        ],
    ) == ["3", "2", "1", "1"]

    assert s.tokenize("", ["a:1"]) == []
    assert s.tokenize("xyz", ["ab:1"]) == ["x", "y", "z"]
    assert s.tokenize("aa", ["a:1", "a:2"]) == ["2", "2"]

    print("string_tokenizer: all tests passed")


if __name__ == "__main__":
    _run_tests()